``uptime``: Support files in uptime

AnirbanHalder654322 opened the pull request:
Fixes #6393

Edit1 : I strace-d the whole thing , we seem to use more than 2x amount of read sys calls (will result in more than 100 extra read calls for very large iterations)compared  to GNU uptime, which i believe is due to ``iter_all_records_from`` function  in uucore::src::lib::features::utmpx

sylvestre commented on Cargo.toml:
```diff
@@ -490,8 +490,16 @@ sha1 = { version = "0.10.6", features = ["std"] }
+  "signals",
+  "utmpx",
+] }
+serde = { version = "1.0.202", features = ["derive"] }
```
do we really need 3 new dependencies ?

AnirbanHalder654322 replied:
We don't need them for building , but i am basically serializing utmp records into bytes and writing it to a file to test the ``uptime`` on a file instead of just shipping binary files into the repo for testing, leading to much more easily auditable code. I could try to write a serializer to turn my struct into binary instead of using dependencies but i believe it would require some research into what byte order different architectures use.

sylvestre replied:
sorry, i missed that it was in [dev-dependencies]
sorry

AnirbanHalder654322 commented:
Changes summary since last review.

1. Testing using file with dynamically created binary utmp records. Not testing for macos , since i don't have access to make to figure out creation of the binaries. Added a todo for anyone who owns a mac.
2.  Added utmpxname() detection since it returns an error in mac systems when the filename doesn't end up 'x'.
3.  Some lint stuff.

AnirbanHalder654322 commented:
Changes since last push:

The test ``test_uptime::test_uptime_for_file_without_utmpx_records`` has been disabled for ``freebsd`` , our ``uucore::utmpx`` module uses ``utmpxname()`` sys call to set the file from which we iterate through, freebsd uses different sets of syscall which function somewhat similarly yet have differences. Comments have been added to the tests which may pass the CI but not really in the intended manner.

AnirbanHalder654322 commented:
The failing CI tests are unrelated to my PR.

sylvestre commented on tests/by-util/test_uptime.rs:
```diff
@@ -22,6 +36,221 @@ fn test_uptime() {
+/// Checks whether uptime displays the correct stderr msg when its called with a fifo
+#[test]
+#[cfg(target_os = "linux")]
+#[ignore = "disabled until fixed"]
```
please create a bug to detail what needs to be fixed 
and add the bug number here :)

AnirbanHalder654322 replied:
Well, i changed the way the test works to fix it instead.

AnirbanHalder654322 commented:
@sylvestre  does this require any more changes ?

sylvestre commented on Cargo.toml:
```diff
@@ -509,6 +510,13 @@ rlimit = "0.10.1"
 
+
+[target.'cfg(all(target_family= "unix",not(target_os = "macos")))'.dev-dependencies]
+serde = { version = "1.0.202", features = ["derive"] }
```
could you please document where they are used ?
and what for

AnirbanHalder654322 replied:
Yea, sure.

sylvestre commented on src/uu/uptime/src/platform/unix.rs:
```diff
@@ -3,49 +3,172 @@
 extern "C" {
     fn GetTickCount() -> uucore::libc::uint32_t;
 }
+quick_error! {
```
please use thiserror

AnirbanHalder654322 replied:
Sure, i made the change

AnirbanHalder654322 commented:
Changes since last push:
- Removed quick_error in favor of thiserror
- Added some documentation on the added dependencies.

sylvestre commented:
thanks!
